hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+			final String dirPath,
+			final long logPeriodMs,
+			final Signer signer,
+			final boolean startWriteAtCompleteWindow,
+			final RecordStreamType streamType
+	) throws NoSuchAlgorithmException {
+		this.dirPath = dirPath;
+		this.logPeriodMs = logPeriodMs;
+		this.signer = signer;
+		this.startWriteAtCompleteWindow = startWriteAtCompleteWindow;
+		this.streamType = streamType;
+		this.streamDigest = MessageDigest.getInstance(DigestType.SHA_384.algorithmName());
+		this.metadataStreamDigest = MessageDigest.getInstance(DigestType.SHA_384.algorithmName());
+	}
+
+	@Override
+	public void addObject(T object) {
+		if (checkIfShouldWriteNewFile(object)) {
+			// if we are currently writing a file,
+			// finish current file and generate signature file
+			closeCurrentAndSign();
+			// write the beginning of the new file
+			beginNew(object);
+		}
+
+		// if recordStreamFile is null, it means startWriteAtCompleteWindow is true,
+		// and we are still in the first incomplete window, so we don't serialize this object;
+		// so we only serialize the object when stream is not null
+		if (recordStreamFileBuilder != null) {
+			consume(object);
+		}
+
+		// update runningHash
+		this.runningHash = object.getRunningHash();
+	}
+
+	/**
+	 * check whether need to start a new file
+	 *
+	 * @param object
+	 * 		the object to be written into file
+	 * @return whether we should start a new File for writing this object
+	 */
+	public boolean checkIfShouldWriteNewFile(T object) {
+		Instant currentConsensusTimestamp = object.getTimestamp();
+		boolean result;
+		if (lastConsensusTimestamp == null && !startWriteAtCompleteWindow) {
+			// this is the first object, we should start writing it to new File
+			result = true;
+		} else if (lastConsensusTimestamp == null) {
+			// this is the first object, we should wait for the first complete window
+			result = false;
+		} else {
+			// if lastConsensusTimestamp and currentConsensusTimestamp are in different periods,
+			// we should start a new file
+			result =
+					getPeriod(lastConsensusTimestamp, logPeriodMs) != getPeriod(currentConsensusTimestamp, logPeriodMs);
+		}
+		// update lastConsensusTimestamp
+		lastConsensusTimestamp = currentConsensusTimestamp;
+		return result;
+	}
+
+	/**
+	 * if recordStreamFile is not null:
+	 * write last runningHash to current file;
+	 * close current file;
+	 * and generate a corresponding signature file
+	 */
+	public void closeCurrentAndSign() {
+		if (recordStreamFileBuilder != null) {
+			// generate file name
+			final var firstTxnTimestamp =
+					recordStreamFileBuilder.getRecordStreamItems(0).getRecord().getConsensusTimestamp();
+			this.file = new File(
+					generateStreamFilePath(
+							Instant.ofEpochSecond(firstTxnTimestamp.getSeconds(), firstTxnTimestamp.getNanos())));
+			this.fileNameShort = file.getName();
+			if (file.exists() && !file.isDirectory()) {
+				LOG.debug(OBJECT_STREAM.getMarker(), "Stream file already exists {}", fileNameShort);
```

Comment:
If it already exists shouldn't it delete it and recreate it? Or seems like a serious error condition.

Reply:
The behavior in this scenario is the same as in the [V5 writer.](https://github.com/hashgraph/swirlds-open-review/blob/a76d4b05ff587a6be035f75c7e4c454283c2d4ca/swirlds-common/src/main/java/com/swirlds/common/stream/TimestampStreamFileWriter.java#L213-L215) Should we look into handling it in a different way?

Reply:
The only way I believe this can happen is if a node has just reconnected and is replaying transactions (because it got a relatively old signed state from the teacher).

In this case, experience has shown the most sensible thing to do is just skip writing the file.

Reply:
Got it, thanks Michael.